### PR TITLE
SPEC-1767 Fix race in CMAP test 'maxConnecting is enforced'

### PR DESCRIPTION
**Pull request by ShaneHarvey**
Tested in python here: https://evergreen.mongodb.com/version/5fc6afbdd1fe0771a322eaec
And a second time for good measure: https://evergreen.mongodb.com/version/5fc6edb032f417146256a3aa

### CODE REVIEW

File: source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml
Comment (patrickfreed) [inline source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml:36]: The waiting here is to ensure thread1 has begun establishing a connection, not that the connection is ready (aka established).

File: source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml
Comment (patrickfreed) [inline source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml:24]: The way the Rust driver runs these tests is that it essentially separates each threads workloads into separate queues and executes them all concurrently, so having the starts happen before the main thread waits will cause all the checkOuts to happen simultaneously. This does seem like an odd way to perform these tests however. I think they're supposed to be run like the main thread dispatching tasks sequentially as they're listed in the format. Is this how Python runs the tests?

Comment (patrickfreed) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml:24]: So it looks like Java, Go, and Node dispatch the threaded tasks in order from the main thread, and Ruby separates them per thread at the beginning and begins execute them one after another once the `start` operation for a given thread is called like Rust does.

I believe the tests were _intended_ to be executed like the former since it makes them much easier to write and read. I'm happy to update the Rust runner to accommodate this, but it is a little worrisome that we'd also make Ruby do the same change. If we think it constitutes a bug in the two implementations, then its probably worth fixing though.

Comment (patrickfreed) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml:24]: After thinking about it some more, I feel like this is a bug in the Ruby and Rust implementations and that drivers should be dispatching threaded tasks from the main thread as opposed to preprocessing them on a per-thread basis, and that we should make this explicit in the spec readme. cc @divjotarora do you agree?

also cc @p-mongo

File: source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml
Comment (ShaneHarvey) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml]: Fixed.

File: source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml
Comment (ShaneHarvey) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml:24]: Yeah I agree that the operations should be dispatched in order.

Comment (divjotarora) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml:24]: Also agreed. I haven't read the spec tests README yet, but it's probably worth updating it to clarify this if we're considering it a bug in those test runners.

**Comment by patrickfreed** [comment]
I updated the rust driver to dispatch tasks in order and pulled in your test, evg patch here: https://evergreen.mongodb.com/version/5fc832813e8e861c6404f5a1

If all goes well, even the windows tasks should all succeed. I'll probably run them a few times if they come back green to make sure.

File: source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml
Comment (patrickfreed) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml:24]: Yep, @ShaneHarvey could you add that as part of this PR then? I think this would be the first test that actually requires this ordering.

**Review by patrickfreed — COMMENTED** [review]
Just restarted my evergreen patch, but the first run seemed to have no maxConnecting failures.

File: source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml
Comment (ShaneHarvey) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml:24]: Updated, PTAL.

**Review by patrickfreed — APPROVED** [review]

File: source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml
Comment (patrickfreed) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-maxConnecting-is-enforced.yml:24]: looks good to me!

**Comment by ShaneHarvey** [comment]
Waiting for @divjotarora 's review before merging.

**Comment by divjotarora** [comment]
@ShaneHarvey I haven't been reviewing the other changes related to DRIVERS-781, so no need to wait for my review.

**Pull request merged**
